### PR TITLE
wait for certs to appear before starting envoy

### LINKED ISSUE #6508: budget exhausted trying to fulfill the desired configuration- egress/ingress gateway
**Describe the bug**

Both IngressGateway and egressGateway are erroring giving the following error..

Permanent error: budget exhausted trying to fulfill the desired configuration**

2018-06-22T03:59:32.874616Z	warn	Epoch 0 terminated with an error: exit status 1
2018-06-22T03:59:32.874799Z	warn	Aborted all epochs
**2018-06-22T03:59:32.874871Z	error	Permanent error: budget exhausted trying to fulfill the desired configuration**
2018-06-22T03:59:32.874919Z	error	cannot start the proxy with the desired configuration 

**Expected behavior**

Expect both ingress and egress gateway to create containers without errors.

**Version**
What version of istio and Kubernetes are you using? Use `istioctl version` and `kubectl version`
Client Version: version.Info{Major:"1", Minor:"9", GitVersion:"v1.9.4", GitCommit:"bee2d1505c4fe820744d26d41ecd3fdd4a3d6546", GitTreeState:"clean", BuildDate:"2018-03-12T16:29:47Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"9", GitVersion:"v1.9.4", GitCommit:"bee2d1505c4fe820744d26d41ecd3fdd4a3d6546", GitTreeState:"clean", BuildDate:"2018-03-12T16:21:35Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"linux/amd64"}
Version: 0.8.0

GitRevision: 6f9f420f0c7119ff4fa6a1966a6f6d89b1b4db84
User: root@48d5ddfd72da
Hub: docker.io/istio
GolangVersion: go1.10.1
BuildStatus: Clean
**Is Istio Auth enabled or not?**
Did you install the stable istio.yaml, istio-auth.yaml.... or if using the Helm chart please provide full command line input.

**Environment**
Azure VM''s

### LINKED ISSUE #9466: Permanent error: budget exhausted trying to fulfill the desired configuration
**K8**: 1.10.7
**istio**: 1.0.2
**mTLS**: enabled
**control plane security**: enabled

Some of the components randomly crash during the start up process. In some instances (like shown below) `pilot` crashed 3 times.

```
istio-citadel-54dc4c6694-t68vp                      1/1       Running   0          10m       10.4.8.101   gke-riro-cluster-control-pool-2-1fe826e8-krvk
istio-egressgateway-685f596699-4w4ct                1/1       Running   1          10m       10.4.9.18    gke-riro-cluster-control-pool-2-1fe826e8-vmcx
istio-galley-ff5d4cbb7-lpfx6                        1/1       Running   0          10m       10.4.9.21    gke-riro-cluster-control-pool-2-1fe826e8-vmcx
istio-ingressgateway-b7b69bc4d-z54cn                1/1       Running   1          10m       10.4.9.17    gke-riro-cluster-control-pool-2-1fe826e8-vmcx
istio-mixer-policy-5cd4c749c9-9xfjg                 2/2       Running   1          10m       10.4.9.19    gke-riro-cluster-control-pool-2-1fe826e8-vmcx
istio-mixer-statsd-prom-bridge-d44f6f584-v882p      1/1       Running   0          10m       10.4.6.108   gke-riro-cluster-control-pool-2-1fe826e8-j2sv
istio-mixer-telemetry-5666864948-wpvhh              2/2       Running   1          10m       10.4.6.107   gke-riro-cluster-control-pool-2-1fe826e8-j2sv
istio-pilot-7f57d6d6f5-ssnvr                        2/2       Running   3          10m       10.4.8.102   gke-riro-cluster-control-pool-2-1fe826e8-krvk
istio-sidecar-7fffd9cbb5-6kbtm                      1/1       Running   0          10m       10.4.9.22    gke-riro-cluster-control-pool-2-1fe826e8-vmcx
```

Im assuming this is mostly due to "some" slowness in `root-cert.pem` generation as it eventually recovers ?

```
2018-10-22T22:42:28.533209Z	warn	Epoch 0 terminated with an error: exit status 1
2018-10-22T22:42:28.533310Z	warn	Aborted all epochs
2018-10-22T22:42:28.533350Z	warn	Failed to delete config file /etc/istio/proxy/envoy-rev0.json for 0, remove /etc/istio/proxy/envoy-rev0.json: no such file or directory
2018-10-22T22:42:28.533368Z	info	Epoch 0: set retry delay to 1m42.4s, budget to 0
2018-10-22T22:44:10.933641Z	info	Reconciling configuration (budget 0)
2018-10-22T22:44:10.933738Z	info	Epoch 0 starting
2018-10-22T22:44:10.933868Z	info	Envoy command: [-c /etc/istio/proxy/envoy.yaml --restart-epoch 0 --drain-time-s 2 --parent-shutdown-time-s 3 --service-cluster istio-pilot --service-node sidecar~10.4.8.102~istio-pilot-7f57d6d6f5-ssnvr.control~control.svc.cluster.local --max-obj-name-len 189 --allow-unknown-fields -l warn]
[2018-10-22 22:44:11.043][56][info][main] external/envoy/source/server/server.cc:190] initializing epoch 0 (hot restart version=10.200.16384.256.options=capacity=16384, num_slots=8209 hash=228984379728933363 size=4882536)
[2018-10-22 22:44:11.043][56][info][main] external/envoy/source/server/server.cc:192] statically linked extensions:
[2018-10-22 22:44:11.043][56][info][main] external/envoy/source/server/server.cc:194]   access_loggers: envoy.file_access_log,envoy.http_grpc_access_log
[2018-10-22 22:44:11.043][56][info][main] external/envoy/source/server/server.cc:197]   filters.http: envoy.buffer,envoy.cors,envoy.ext_authz,envoy.fault,envoy.filters.http.header_to_metadata,envoy.filters.http.jwt_authn,envoy.filters.http.rbac,envoy.grpc_http1_bridge,envoy.grpc_json_transcoder,envoy.grpc_web,envoy.gzip,envoy.health_check,envoy.http_dynamo_filter,envoy.ip_tagging,envoy.lua,envoy.rate_limit,envoy.router,envoy.squash,istio_authn,jwt-auth,mixer
[2018-10-22 22:44:11.043][56][info][main] external/envoy/source/server/server.cc:200]   filters.listener: envoy.listener.original_dst,envoy.listener.proxy_protocol,envoy.listener.tls_inspector
[2018-10-22 22:44:11.043][56][info][main] external/envoy/source/server/server.cc:203]   filters.network: envoy.client_ssl_auth,envoy.echo,envoy.ext_authz,envoy.filters.network.rbac,envoy.filters.network.thrift_proxy,envoy.http_connection_manager,envoy.mongo_proxy,envoy.ratelimit,envoy.redis_proxy,envoy.tcp_proxy,mixer
[2018-10-22 22:44:11.043][56][info][main] external/envoy/source/server/server.cc:205]   stat_sinks: envoy.dog_statsd,envoy.metrics_service,envoy.stat_sinks.hystrix,envoy.statsd
[2018-10-22 22:44:11.043][56][info][main] external/envoy/source/server/server.cc:207]   tracers: envoy.dynamic.ot,envoy.lightstep,envoy.zipkin
[2018-10-22 22:44:11.043][56][info][main] external/envoy/source/server/server.cc:210]   transport_sockets.downstream: alts,envoy.transport_sockets.capture,raw_buffer,tls
[2018-10-22 22:44:11.043][56][info][main] external/envoy/source/server/server.cc:213]   transport_sockets.upstream: alts,envoy.transport_sockets.capture,raw_buffer,tls
[2018-10-22 22:44:11.127][56][info][config] external/envoy/source/server/configuration_impl.cc:50] loading 0 static secret(s)
[2018-10-22 22:44:11.130][56][critical][main] external/envoy/source/server/server.cc:78] error initializing configuration '/etc/istio/proxy/envoy.yaml': unable to read file: /etc/certs/root-cert.pem
[2018-10-22 22:44:11.130][56][info][main] external/envoy/source/server/server.cc:472] exiting
2018-10-22T22:44:11.132544Z	warn	Epoch 0 terminated with an error: exit status 1
2018-10-22T22:44:11.132584Z	warn	Aborted all epochs
2018-10-22T22:44:11.132621Z	warn	Failed to delete config file /etc/istio/proxy/envoy-rev0.json for 0, remove /etc/istio/proxy/envoy-rev0.json: no such file or directory
2018-10-22T22:44:11.132626Z	error	Permanent error: budget exhausted trying to fulfill the desired configuration
2018-10-22T22:44:11.132630Z	error	cannot start the proxy with the desired configuration
2018-10-22T22:44:11.132641Z	error
```

I dont see any errors in citadel either.

### LINKED ISSUE #9659: Sidecar doesn't start since certificates are missing on the sidecar pod.
**Describe the bug**
We have a scenario, where the egress gateway originates TLS with certificates that are mounted to the path `/etc/istio/egressgateway-certs/`. 
This path is automatically mounted, if a secret with the name istio-egressgateway-certs exists.

We configure the the TLS origination by a destination rule:
```
piVersion: networking.istio.io/v1alpha3
kind: DestinationRule
metadata:
  name: egressgateway-pinger
spec:
  host: pinger.services.acme.com
  subsets:
  - name: pinger
    trafficPolicy:
      portLevelSettings:
      - port:
          number: 9000
        tls:
          caCertificates: /etc/istio/egressgateway-certs/ca.crt
          clientCertificate: /etc/istio/egressgateway-certs/client.crt
          mode: MUTUAL
          privateKey: /etc/istio/egressgateway-certs/client.key
          sni: pinger.services.acme.com
```

Our pods have sidecars, that do not start since the sidecar expects to find certificates in
`/etc/istio/egressgateway-certs/`  

I extracted parts of the sidecar envoy log:

```
2018-10-30T15:33:14.237072Z	info	Monitored certs: []envoy.CertSource{envoy.CertSource{Directory:"/etc/certs/", Files:[]string{"cert-chain.pem", "key.pem", "root-cert.pem"}}}
2018-10-30T15:33:14.237222Z	info	Opening status port 15020

2018-10-30T15:33:14.237236Z	info	Starting proxy agent
2018-10-30T15:33:14.237487Z	info	Received new config, resetting budget
2018-10-30T15:33:14.237497Z	info	Reconciling retry (budget 10)
2018-10-30T15:33:14.237525Z	info	Epoch 0 starting
2018-10-30T15:33:14.238236Z	info	Envoy command: [-c /etc/istio/proxy/envoy-rev0.json --restart-epoch 0 --drain-time-s 45 --parent-shutdown-time-s 60 --service-cluster client.default --service-node sidecar~100.96.1.17~client-845b478b6d-lkwgd.default~default.svc.cluster.local --max-obj-name-len 189 --allow-unknown-fields -l warning --v2-config-only]
[2018-10-30 15:33:14.270][16][warning][upstream] external/envoy/source/common/config/grpc_mux_impl.cc:243] gRPC config stream closed: 14, no healthy upstream
[2018-10-30 15:33:14.271][16][warning][upstream] external/envoy/source/common/config/grpc_mux_impl.cc:43] Unable to establish new stream
[2018-10-30 15:33:14.284][16][warning][upstream] external/envoy/source/common/config/grpc_mux_impl.cc:243] gRPC config stream closed: 14, no healthy upstream
[2018-10-30 15:33:14.284][16][warning][upstream] external/envoy/source/common/config/grpc_mux_impl.cc:43] Unable to establish new stream
2018-10-30T15:33:15.929821Z	info	Envoy proxy is NOT ready: 2 errors occurred:

* envoy stat missing: cluster_manager.cds.update_success
* envoy stat missing: listener_manager.lds.update_success
2018-10-30T15:33:17.929777Z	info	Envoy proxy is NOT ready: 2 errors occurred:

* envoy stat missing: cluster_manager.cds.update_success
* envoy stat missing: listener_manager.lds.update_success
[2018-10-30 15:33:19.366][16][warning][upstream] external/envoy/source/common/config/grpc_mux_impl.cc:226] gRPC config for type.googleapis.com/envoy.api.v2.Cluster update rejected: unable to read file: /etc/istio/egressgateway-certs/client.crt
[2018-10-30 15:33:19.366][16][warning][config] bazel-out/k8-opt/bin/external/envoy/source/common/config/_virtual_includes/grpc_mux_subscription_lib/common/config/grpc_mux_subscription_impl.h:70] gRPC config for type.googleapis.com/envoy.api.v2.Cluster rejected: unable to read file: /etc/istio/egressgateway-certs/client.crt
[2018-10-30 15:33:19.459][16][warning][config] src/envoy/utils/mixer_control.cc:171] ExtractInfo  metadata missing:
[2018-10-30 15:33:19.461][16][warning][config] src/envoy/utils/mixer_control.cc:171] ExtractInfo  metadata missing:
...
2018-10-30 15:33:19.529][25][warning][config] src/envoy/utils/mixer_control.cc:171] ExtractInfo  metadata missing:
[2018-10-30 15:33:19.529][25][warning][config] src/envoy/utils/mixer_control.cc:171] ExtractInfo  metadata missing:
[2018-10-30 15:33:19.532][25][warning][config] src/envoy/utils/mixer_control.cc:171] ExtractInfo  metadata missing:
[2018-10-30 15:33:19.532][25][warning][config] src/envoy/utils/mixer_control.cc:171] ExtractInfo  metadata missing:
[2018-10-30 15:33:19.532][25][warning][config] src/envoy/utils/mixer_control.cc:171] ExtractInfo  metadata missing:
[2018-10-30 15:33:19.532][25][warning][config] src/envoy/utils/mixer_control.cc:171] ExtractInfo  metadata missing:
...
2018-10-30T15:33:19.935145Z	info	Envoy proxy is NOT ready: 1 error occurred:

* envoy stat missing: cluster_manager.cds.update_success
2018-10-30T15:33:21.935445Z	info	Envoy proxy is NOT ready: 1 error occurred:

* envoy stat missing: cluster_manager.cds.update_success
2018-10-30T15:33:23.935120Z	info	Envoy proxy is NOT ready: 1 error occurred:

* envoy stat missing: cluster_manager.cds.update_success
2018-10-30T15:33:25.934052Z	info	Envoy proxy is NOT ready: 1 error occurred:
...
```

**Expected behavior**
We would expect, that the sidecar ignores the missing certificates, as it doesn't need them.
This seems to be the behaviour in 1.0.
Or some method to specify the gateway, where the destination rule should be active.

**Steps to reproduce the bug**
* Create a secret with the required name for egress certificates.
* Deploy a Destination rule like in the example above.
* Start a pod with a sidecar. 

**Version**
Istio 1.1.0-snapshot.2 - October 2018 Release
and the current master.

### LINKED ISSUE #10205: "Failed to start secure grpc server open /etc/certs/cert-chain.pem" on Pilot
**Describe the bug**
Similar to #9466 I'm observing Pilot crashing couple of times when it waits for citadel to generate the certs. It eventually recovers but would like to get to the bottom of why it doesnt start cleanly. This may or may not be related to #9466 hence raising this issue.
```
gc 1 @0.032s 4%: 0.015+2.2+1.3 ms clock, 0.031+0.33/0.28/1.8+2.7 ms cpu, 4->4->2 MB, 5 MB goal, 2 P
gc 2 @0.049s 7%: 0.011+2.8+1.8 ms clock, 0.023+0.38/0.20/2.6+3.6 ms cpu, 4->4->2 MB, 5 MB goal, 2 P
2018-11-30T05:39:20.460622Z	info	FLAG: --appNamespace=""
2018-11-30T05:39:20.460810Z	info	FLAG: --caCertFile="/etc/istio/certs/root-cert.pem"
2018-11-30T05:39:20.460826Z	info	FLAG: --certFile="/etc/istio/certs/cert-chain.pem"
2018-11-30T05:39:20.460934Z	info	FLAG: --clusterRegistriesNamespace=""
2018-11-30T05:39:20.460951Z	info	FLAG: --configDir=""
2018-11-30T05:39:20.461019Z	info	FLAG: --consulserverInterval="2s"
2018-11-30T05:39:20.461031Z	info	FLAG: --consulserverURL=""
2018-11-30T05:39:20.461067Z	info	FLAG: --ctrlz_address="127.0.0.1"
2018-11-30T05:39:20.461134Z	info	FLAG: --ctrlz_port="9876"
2018-11-30T05:39:20.461148Z	info	FLAG: --disable-install-crds="false"
2018-11-30T05:39:20.461207Z	info	FLAG: --discoveryCache="true"
2018-11-30T05:39:20.461218Z	info	FLAG: --domain="cluster.local"
2018-11-30T05:39:20.461223Z	info	FLAG: --grpcAddr=":15010"
2018-11-30T05:39:20.461303Z	info	FLAG: --help="false"
2018-11-30T05:39:20.461314Z	info	FLAG: --httpAddr=":8080"
2018-11-30T05:39:20.461319Z	info	FLAG: --keyFile="/etc/istio/certs/key.pem"
2018-11-30T05:39:20.461380Z	info	FLAG: --kubeconfig=""
2018-11-30T05:39:20.461384Z	info	FLAG: --log_as_json="false"
2018-11-30T05:39:20.461457Z	info	FLAG: --log_caller=""
2018-11-30T05:39:20.461468Z	info	FLAG: --log_output_level="default:info"
2018-11-30T05:39:20.461505Z	info	FLAG: --log_rotate=""
2018-11-30T05:39:20.461564Z	info	FLAG: --log_rotate_max_age="30"
2018-11-30T05:39:20.461575Z	info	FLAG: --log_rotate_max_backups="1000"
2018-11-30T05:39:20.461635Z	info	FLAG: --log_rotate_max_size="104857600"
2018-11-30T05:39:20.461646Z	info	FLAG: --log_stacktrace_level="default:none"
2018-11-30T05:39:20.461688Z	info	FLAG: --log_target="[stdout]"
2018-11-30T05:39:20.461749Z	info	FLAG: --mcpMaxMsgSize="4194304"
2018-11-30T05:39:20.461826Z	info	FLAG: --mcpServerAddrs="[]"
2018-11-30T05:39:20.461838Z	info	FLAG: --meshConfig="/etc/istio/config/mesh"
2018-11-30T05:39:20.461914Z	info	FLAG: --monitoringAddr=":9093"
2018-11-30T05:39:20.461980Z	info	FLAG: --namespace=""
2018-11-30T05:39:20.461991Z	info	FLAG: --networksConfig="/etc/istio/config/meshNetworks"
2018-11-30T05:39:20.462034Z	info	FLAG: --plugins="[authn,authz,health,mixer,envoyfilter]"
2018-11-30T05:39:20.462095Z	info	FLAG: --profile="true"
2018-11-30T05:39:20.462111Z	info	FLAG: --registries="[Kubernetes]"
2018-11-30T05:39:20.462172Z	info	FLAG: --resync="1m0s"
2018-11-30T05:39:20.462210Z	info	FLAG: --secureGrpcAddr=":15012"
2018-11-30T05:39:20.466921Z	info	mesh configuration (*v1alpha1.MeshConfig)(0xc4202665a0)(mixer_check_server:"istio-policy.control.svc.cluster.local:15004" mixer_report_server:"istio-telemetry.control.svc.cluster.local:15004" disable_policy_checks:true proxy_listen_port:15001 connect_timeout:<seconds:1 > ingress_class:"istio" ingress_controller_mode:STRICT enable_tracing:true access_log_file:"/dev/stdout" default_config:<config_path:"/etc/istio/proxy" binary_path:"/usr/local/bin/envoy" service_cluster:"istio-proxy" drain_duration:<seconds:45 > parent_shutdown_duration:<seconds:60 > discovery_address:"istio-pilot.control:15011" connect_timeout:<seconds:10 > proxy_admin_port:15000 control_plane_auth_policy:MUTUAL_TLS stat_name_length:189 tracing:<zipkin:<address:"jg-jaeger.monitoring.svc.cluster.local:9411" > > > )

2018-11-30T05:39:20.466953Z	info	version root@e34a3c8d4731-docker.io/istio-1.1.0-snapshot.3-e3653d5a8244c2f30e2e0600cb5af956a575bae1-dirty-Modified
2018-11-30T05:39:20.467483Z	info	flags (*bootstrap.PilotArgs)(0xc420414380)({
 DiscoveryOptions: (envoy.DiscoveryServiceOptions) {
  HTTPAddr: (string) (len=5) ":8080",
  GrpcAddr: (string) (len=6) ":15010",
  SecureGrpcAddr: (string) (len=6) ":15012",
  MonitoringAddr: (string) (len=5) ":9093",
  EnableProfiling: (bool) true,
  EnableCaching: (bool) true
 },
 Namespace: (string) (len=7) "control",
 Mesh: (bootstrap.MeshArgs) {
  ConfigFile: (string) (len=22) "/etc/istio/config/mesh",
  MixerAddress: (string) "",
  RdsRefreshDelay: (*types.Duration)(<nil>)
 },
 Config: (bootstrap.ConfigArgs) {
  ClusterRegistriesNamespace: (string) (len=7) "control",
  KubeConfig: (string) "",
  ControllerOptions: (kube.ControllerOptions) {
   WatchedNamespace: (string) "",
   ResyncPeriod: (time.Duration) 1m0s,
   DomainSuffix: (string) (len=13) "cluster.local",
   ClusterID: (string) "",
   XDSUpdater: (model.XDSUpdater) <nil>,
   stop: (chan struct {}) <nil>
  },
  FileDir: (string) "",
  DisableInstallCRDs: (bool) false,
  Controller: (model.ConfigStoreCache) <nil>
 },
 Service: (bootstrap.ServiceArgs) {
  Registries: ([]string) (len=1 cap=1) {
   (string) (len=10) "Kubernetes"
  },
  Consul: (bootstrap.ConsulArgs) {
   Config: (string) "",
   ServerURL: (string) "",
   Interval: (time.Duration) 2s
  }
 },
 MeshConfig: (*v1alpha1.MeshConfig)(<nil>),
 NetworksConfigFile: (string) (len=30) "/etc/istio/config/meshNetworks",
 CtrlZOptions: (*ctrlz.Options)(0xc4205512c0)({
  Port: (uint16) 9876,
  Address: (string) (len=9) "127.0.0.1"
 }),
 Plugins: ([]string) (len=5 cap=5) {
  (string) (len=5) "authn",
  (string) (len=5) "authz",
  (string) (len=6) "health",
  (string) (len=5) "mixer",
  (string) (len=11) "envoyfilter"
 },
 MCPServerAddrs: ([]string) {
 },
 MCPCredentialOptions: (*creds.Options)(0xc420567c20)({
  CertificateFile: (string) (len=31) "/etc/istio/certs/cert-chain.pem",
  KeyFile: (string) (len=24) "/etc/istio/certs/key.pem",
  CACertificateFile: (string) (len=30) "/etc/istio/certs/root-cert.pem"
 }),
 MCPMaxMessageSize: (int) 4194304
})

2018-11-30T05:39:20.468031Z	info	mesh networks configuration (*v1alpha1.MeshNetworks)(0xc42000c6c8)()

2018-11-30T05:39:20.469982Z	info	registering for apiVersion authentication.istio.io/v1alpha1
gc 3 @0.077s 10%: 0.009+3.1+4.4 ms clock, 0.019+0.24/1.1/2.8+8.9 ms cpu, 5->5->3 MB, 6 MB goal, 2 P
2018-11-30T05:39:20.487898Z	info	registering for apiVersion rbac.istio.io/v1alpha1
2018-11-30T05:39:20.498523Z	info	registering for apiVersion networking.istio.io/v1alpha3
2018-11-30T05:39:20.510761Z	info	registering for apiVersion config.istio.io/v1alpha2
2018-11-30T05:39:20.519277Z	info	CRD controller watching namespaces ""
2018-11-30T05:39:20.519743Z	info	Ingress controller watching namespaces ""
2018-11-30T05:39:20.531004Z	info	Adding Kubernetes registry adapter
2018-11-30T05:39:20.531035Z	info	Primary Cluster name: Kubernetes
2018-11-30T05:39:20.531045Z	info	Service controller watching namespace "" for services, endpoints, nodes and pods, refresh 1m0s
2018-11-30T05:39:20.531462Z	info	ads	Starting ADS server with rateLimiter=10 burst=100
2018-11-30T05:39:20.531698Z	info	Setting up event handlers
2018-11-30T05:39:20.531798Z	info	Starting Secrets controller
2018-11-30T05:39:20.531907Z	info	Waiting for informer caches to sync
2018-11-30T05:39:20.536957Z	info	Discovery service started at http=[::]:8080 grpc=[::]:15010 secure grpc=[::]:15012
2018-11-30T05:39:20.537516Z	info	ControlZ available at 10.12.1.88:9876
2018-11-30T05:39:20.538510Z	info	attempting to acquire leader lease  control/istio-ingress-controller-leader-istio...
2018-11-30T05:39:20.539373Z	error	Failed to start secure grpc server open /etc/certs/cert-chain.pem: no such file or directory
```

**Expected behavior**
Pilot to start without crashing

**Steps to reproduce the bug**
Just following the stock helm deployment would lead to this problem

**Version**
**Istio**: Istio 1.1.0-snapshot.3
**GKE**: 1.11.2-gke.18

**Installation**
Stock helm deployment with mTLS and controlPlaneSecurityEnabled.

**Environment**
GKE

### PR DESCRIPTION
**Pull request by elevran**
This PR waits (some time) for certificates to appear before starting envoy.

There are several open issues (#6508, #9466, #9659 and #10205) reporting Envoy restarting due to missing certs:
`[2018-10-22 22:44:11.130][56][critical][main] external/envoy/source/server/server.cc:78] error initializing configuration '/etc/istio/proxy/envoy.yaml': unable to read file: /etc/certs/root-cert.pem` (from issue #9466).

Envoy is automatically restarted and eventually the certificates are available, configuration is valid and Envoy loads successfully.

The root cause seems to be high delay in having certificates mounted. The process is as follows:
- t0: Pod is started with a new service account. Since certs volume for `/etc/certs` is marked `optional: true`, there is no need to wait for the secret before starting containers.
- t1: Citadel notices the new SA and creates a secret for it.
- t2: Kubelet periodically scans for secrets and populates the mount
- t3: envoy reads the secret

If t3-t0 is too high, envoy will terminate. Citadel time (t1) depends on current load, Kubelet sync time is not under our control (see below, defaults could be as high as 1 min). This PR avoid some of the restarts by testing for certificate existence inside pilot-agent before attempting to start Envoy. Waiting is only enabled when certificates are definitely required (i.e., when control plane authentication is enabled). For other cases, it falls back to the existing behavior. Additionally, pilot-agent only waits up to 2m before giving up and letting Envoy run its course (again - falling back to current behavior). Informational log messages are displayed with exponential backoff. Note that only logging is throttled, checking continues on much shorter period and would terminate as soon as certificates are available.

Fixes #6508 Fixes #9466 Fixes #9659 Fixes #10205.

For reference:
- https://kubernetes.io/docs/concepts/configuration/secret/#secret-and-pod-lifetime-interaction:
> When a pod is created via the API, there is no check whether a referenced secret exists. Once a pod is scheduled, the kubelet will try to fetch the secret value. If the secret cannot be fetched because it does not exist or because of a temporary lack of connection to the API server, kubelet will periodically retry.

- https://kubernetes.io/docs/concepts/configuration/secret/:
> Kubelet is checking whether the mounted secret is fresh on every periodic sync. However, it is using its local cache for getting the current value of the Secret. The type of the cache is configurable using the (ConfigMapAndSecretChangeDetectionStrategy field in KubeletConfiguration struct)... total delay from the moment when the Secret is updated to the moment when new keys are projected to the Pod can be as long as kubelet sync period + cache propagation delay

### CODE REVIEW

File: pilot/cmd/pilot-agent/main.go [hunk 1]
```diff
@@ -483,6 +488,35 @@ func init() {
...
+	nextLog := time.Now().Add(logDelay)
+	endWait := time.Now().Add(maxWait)
+
+	for {
+		_, err := os.Stat(fname)
+		if err == nil {
+			return
+		}
+		if !os.IsNotExist(err) { // another error (e.g., permission) - likely no point in waiting longer
+			log.Infof("error while waiting for certificates: %s", err.Error())
```
Comment (ymesika) [inline pilot/cmd/pilot-agent/main.go]: Maybe `log.Errorf` is better level here

Comment (ymesika) [reply pilot/cmd/pilot-agent/main.go]: Or better - `log.Errora` with string and `err` as an argument

File: pilot/cmd/pilot-agent/main.go [hunk 2]
```diff
@@ -256,6 +257,10 @@ var (
 			}
 
 			log.Infof("Monitored certs: %#v", certs)
+			// since Envoy needs the certs for mTLS, we wait for them to become available before starting it
+			if controlPlaneAuthPolicy == meshconfig.AuthenticationPolicy_MUTUAL_TLS.String() {
+				waitForCerts(path.Join(certs[0].Directory, certs[0].Files[0]), 2*time.Minute)
```
Comment (ymesika) [inline pilot/cmd/pilot-agent/main.go]: Why only on first element of the slice? Maybe loop through all to make sure all certs are avail?

File: pilot/cmd/pilot-agent/main.go [hunk 3]
```diff
@@ -483,6 +488,35 @@ func init() {
...
+		}
+		if !os.IsNotExist(err) { // another error (e.g., permission) - likely no point in waiting longer
+			log.Infof("error while waiting for certificates: %s", err.Error())
+			return
+		}
+
+		now := time.Now()
+		if now.After(endWait) {
+			break
+		} else if now.After(nextLog) {
```
Comment (ymesika) [inline pilot/cmd/pilot-agent/main.go]: else not needed

File: pilot/cmd/pilot-agent/main.go (hunk 2 above)
Comment (ymesika) [reply pilot/cmd/pilot-agent/main.go]: Also, I don't understand what happens if certs aren't available after the max time? It just continues?

Comment (elevran) [reply pilot/cmd/pilot-agent/main.go]: If certs aren't available, it falls back to the current behavior (letting Envoy try and possibly restart). There isn't that much we can do and continuing forever does not add value.

The assumption is that once one certificate file is ready, so are the others. If this is too lax, I can certainly loop through `certs` array and check first file from each. I believe all certs in a directory are coming from the same secret (i,e., cert, key and chain are all parts of one secret)

File: pilot/cmd/pilot-agent/main.go (hunk 3 above)
Comment (elevran) [reply pilot/cmd/pilot-agent/main.go]: removed

File: pilot/cmd/pilot-agent/main.go (hunk 2 above)
Comment (ymesika) [reply pilot/cmd/pilot-agent/main.go]: > If certs aren't available, it falls back to the current behavior (letting Envoy try and possibly restart). There isn't that much we can do and continuing forever does not add value.

Fair enough but maybe print some summarising error message?

File: pilot/cmd/pilot-agent/main.go [hunk 4]
```diff
@@ -483,6 +488,36 @@ func init() {
...
+	nextLog := time.Now().Add(logDelay)
+	endWait := time.Now().Add(maxWait)
+
+	for {
+		_, err := os.Stat(fname)
+		if err == nil {
+			return
+		}
+		if !os.IsNotExist(err) { // another error (e.g., permission) - likely no point in waiting longer
+			log.Errora("error while waiting for certificates:", err.Error())
```
Comment (ymesika) [inline pilot/cmd/pilot-agent/main.go]: I meant `log.Errora("error while waiting for certificates", err)` but that's fine too

Comment (elevran) [reply pilot/cmd/pilot-agent/main.go]: changed

**Pull request merged**